### PR TITLE
bug(stages): TxLookup/Sender stages fix range if there is no tx

### PR DESCRIPTION
**Pull request by rakita**
While testing stage sync on very small ranges there is a possibility that if the next block range does not contain transactions `RangeInclusive` would take one transaction from the future block and when that future blocks come to execution it will crash the node as its transaction was already inserted.

Added `reth db drop` support for TxLookup

and Optimized fetching of tx number, previously we iterated over all entried of `BodyIndices` table, and this was not needed.

### CODE REVIEW

**Review by onbjerg — CHANGES_REQUESTED** [review]
This fundamentally changes how the stage commits. Currently it commits based on a target number of transactions, this PR replaces that with a block based threshold

Edit: Unless `target_block` is always the tip?

**Comment by rakita** [comment]
> This fundamentally changes how the stage commits. Currently it commits based on a target number of transactions, this PR replaces that with a block based threshold
> 
> Edit: Unless `target_block` is always the tip?

It doesn't change behaviour (or it shouldn't), it fixes the bug and optimizes things.

see this line: https://github.com/paradigmxyz/reth/pull/3479/files#diff-bfe079ce420f41bfe72f6c4ae6507a6d4647b664af2f154fdf5372f38b9c5146L93

**Review by onbjerg — COMMENTED** [review]
I'll let someone else review this

**Comment by rakita** [comment]
Tested this on my branch revm state merge branch.

File: bin/reth/src/stage/drop.rs [hunk 1]
```diff
@@ -152,6 +152,14 @@ impl Command {
                     )?;
                     insert_genesis_header::<DatabaseEnv>(tx, self.chain)?;
                 }
+                StageEnum::TxLookup => {
```
Comment (gakonst) [inline bin/reth/src/stage/drop.rs:155]: :+1:

File: crates/stages/src/test_utils/test_db.rs [hunk 2]
```diff
@@ -246,13 +246,19 @@ impl TestTransaction {
...
+                    first_tx_num: next_tx_num,
+                    tx_count: block.body.len() as u64,
+                };
+
+                if !block.body.is_empty() {
+                    tx.put::<tables::TransactionBlock>(
+                        block_body_indices.last_tx_num(),
+                        block.number,
+                    )?;
+                }
```
Comment (gakonst) [inline crates/stages/src/test_utils/test_db.rs:259]: ah this is required due to querying the block of the tx by number in the code

File: crates/stages/src/stage.rs [hunk 3]
```diff
@@ -77,30 +77,46 @@ impl ExecInput {
         &self,
         provider: &DatabaseProviderRW<'_, DB>,
         tx_threshold: u64,
-    ) -> Result<(RangeInclusive<TxNumber>, RangeInclusive<BlockNumber>, bool), StageError> {
+    ) -> Result<(Range<TxNumber>, RangeInclusive<BlockNumber>, bool), StageError> {
```
Comment (gakonst) [inline crates/stages/src/stage.rs:80]: we discussed before but should we consider going to `Range` instead of `RangeInclusive` everywhere to be consistent? recall you didnt love it, also recall me and @onbjerg not loving it at the time..no storng opinions just askign the q

File: crates/stages/src/stage.rs [hunk 4]
```diff
@@ -77,30 +77,46 @@ impl ExecInput {
...
+            // get tx block number
+            let end_block_number =
+                provider.transaction_block(first_tx_num + tx_cnt)?.expect("block of tx must exist");
+            // we want to get range of all transactions of this block, so we are fetching block
+            // body.
+            let end_block_body = provider
+                .block_body_indices(end_block_number)?
+                .ok_or(ProviderError::BlockBodyIndicesNotFound(target_block))?;
+            (end_block_number, end_block_body.next_tx_num() - first_tx_num)
+        };
```
Comment (gakonst) [inline crates/stages/src/stage.rs:116]: i like how this gets around the iteratioon by querying `transaction_block`.

**Review by gakonst — APPROVED** [review]
Seems OK with me but the lack of unit tests for this function makes me scared to merge. cc @rkrasiuk for another pass

File: crates/stages/src/stage.rs (hunk 4 above)
Comment (rakita) [reply crates/stages/src/stage.rs:116]: It is edge case that I mitigate it with `transaction_block`

File: crates/stages/src/test_utils/test_db.rs (hunk 2 above)
Comment (rakita) [reply crates/stages/src/test_utils/test_db.rs:259]: Yep it fails the tests if not set.

File: crates/stages/src/stage.rs (hunk 3 above)
Comment (rakita) [reply crates/stages/src/stage.rs:80]: As I mostly fixed a lot of bugs with this, imo i feel like `RangeInclusive` for blocks is less error prone and same for `Range` for transactions. This was a reason why i used it as is, but in the end it is not a big of a deal having it one or another, and i wouldn't mind a lot if it gets changed.

File: crates/stages/src/stage.rs [hunk 5]
```diff
@@ -77,30 +77,46 @@ impl ExecInput {
...
+
+        let tx_cnt: u64 = min(all_tx_cnt, tx_threshold);
+        let is_final_range = tx_cnt == all_tx_cnt;
+        // get block of this tx
+        let (end_block, tx_cnt) = if is_final_range {
+            (target_block, tx_cnt)
+        } else {
+            // get tx block number
+            let end_block_number =
+                provider.transaction_block(first_tx_num + tx_cnt)?.expect("block of tx must exist");
```
Comment (rkrasiuk) [inline crates/stages/src/stage.rs:109]: in case of `all_tx_cnt < tx_threshold`, we would be querying for `first_tx_num + all_tx_cnt` -> `first_tx_num + target_block_body.next_tx_num() - first_tx_num` -> `target_block_body.next_tx_num()`. seems like a bug, since it isn't guaranteed to exist

File: crates/stages/src/stage.rs [hunk 6]
```diff
@@ -77,30 +77,46 @@ impl ExecInput {
...
+            // if there is no more transaction return back.
+            return Ok((first_tx_num..first_tx_num, start_block..=target_block, true))
         }
-        let is_final_range = end_block_number >= target_block;
-        Ok((first_tx_number..=last_tx_number, start_block..=end_block_number, is_final_range))
+
+        let tx_cnt: u64 = min(all_tx_cnt, tx_threshold);
+        let is_final_range = tx_cnt == all_tx_cnt;
+        // get block of this tx
+        let (end_block, tx_cnt) = if is_final_range {
```
Comment (rkrasiuk) [inline crates/stages/src/stage.rs:104]: imo cleaner to return `(end_block, last_tx_num)` here instead of `tx_cnt`

**Review by rkrasiuk — CHANGES_REQUESTED** [review]
the original bug makes sense, nice find! leaving some comments

File: crates/stages/src/stage.rs (hunk 5 above)
Comment (rakita) [reply crates/stages/src/stage.rs:109]: In this case, we have `let is_final_range = tx_cnt == all_tx_cnt;` and we would not fetch values from db.

Comment (rakita) [reply crates/stages/src/stage.rs:109]: Yeah, only for `all_tx_cnt > tx_threshold` we would need to fetch the block body to get the last transaction of that block.

File: crates/stages/src/stage.rs (hunk 3 above)
Comment (rakita) [reply crates/stages/src/stage.rs:80]: I think most confusion stems not from `Range` `RangeInclusive` but from edge cases when the block does not have transactions but it does have an index on the next block transaction. 
 https://github.com/paradigmxyz/reth/blob/6cbb639f6b4f32ee9908007b64b0a61ea50879d9/crates/storage/db/src/tables/models/blocks.rs#L35-L50

**Review by gakonst — APPROVED** [review]
let's send it - synced w roman
